README.md
firewireDCAM
============

EPICS areaDetector driver for firewire IEEE1394 cameras using libdc1394.

This areaDetector driver was previously published and hosted here: http://controls.diamond.ac.uk/downloads/support/firewireDCAM/ up until release 1-8.

See the release notes [here](RELEASE_NOTES.md)

Introduction
============

This module is a Linux firewire (IEEE 1394) camera driver plug-in for another EPICS support module/framework: http://cars9.uchicago.edu/software/epics/areaDetector.html

Please note that for new applications, the author recommends going for the faster, more versatile and often cheaper GigE-enabled cameras. See the <a href="http://controls.diamond.ac.uk/downloads/support/aravisGigE">aravisGigE areaDetector driver.</a>

Firewire IEEE1394 Cameras
=========================
This module uses a set of open source libraries to control the cameras. The main library is <a href="http://damien.douxchamps.net/ieee1394/libdc1394">libdc1394</a>

libdc1394 is able to control firewire cameras that comply with the IIDC DCAM protocol. Please see the libdc1394 <a href="http://damien.douxchamps.net/ieee1394/cameras/">list of cameras</a> and the <a href="http://damien.douxchamps.net/ieee1394/libdc1394/v2.x/faq/">libdc1394 FAQ</a> for more details about which cameras can be used with libdc1394 (note that the developer of this module has only tested with a limited number of camera models: Point Grey Flea2 and AVR Pike)

Digital video recorders with tapes or other local memory typically does not comply with the IIDC DCAM protocol! Cameras that produce compressed images are not supported either. This module does not and is not planned to provide any compression or decompression features. If such feature is needed it is recommended to develop a compression plug-in to the areaDetector framework.

The module supports both 1394A [400Mb/s] and 1394B [800Mb/s] mode cameras.

Where to get it
---------------
Released (tagged) versions of the module can be downloaded as source code tarballs from areaDetector/firewireDCAM repository.

If you want to contribute code to the repository, please use the usual github forking workflow and issue pull requests with your developments.


Build Instructions
==================
These build instructions assume some knowledge of building EPICS modules. It is a requirement that <a href="http://www.aps.anl.gov/epics/base/index.php">EPICS base</a> has already been downloaded and successfully build with all the common environment variables set. 

Dependencies
------------

**EPICS base and modules**

* EPICS base <a href="http://www.aps.anl.gov/epics/base/R3-14/12.php">R3.14.12.3</a>
* <a href="http://www.aps.anl.gov/epics/modules/soft/asyn">asyn</a> 4-21
* <a href="http://cars9.uchicago.edu/software/epics/areaDetector.html">areaDetector</a> 1-9

Optional Dependencies
---------------------
The <a href="http://controls.diamond.ac.uk/downloads/support/ffmpegServer">ffmpegServer</a> module is recommended for new firewire (or any areaDetector) IOCs. This plugin provide mjpeg compression and with the <a href="http://controls.diamond.ac.uk/downloads/support/nullhttpd">nullhttpd</a> module also http streaming for client side image viewing. It also provides a QT/OpenGL based viewer application.

Firewire IIDC DCAM libraries
----------------------------
The only required Linux firewire camera support library is now: <a href="http://sourceforge.net/project/showfiles.php?group_id=8157">libdc1394</a> 2.2.1
libraw1394 is no longer required.

Kernel Modules
--------------
The libdc1394 libraries depend on a set of drivers/kernel modules that will need to be loaded. There are basically two options when selecting the firewire kernel modules: the traditional ohci1394 (and friends) or the newer 'Juju' stack. Most distributions now ship with the Juju firewire stack. The Juju stack removes the requirement to use the raw1394 kernel module.

When using the Juju firewire stack the following command can inform if the correct kernel modules have been loaded:
<pre>
/sbin/lsmod | grep firewire
firewire_ohci          24695  0 
firewire_core          50151  1 firewire_ohci
crc_itu_t               1717  1 firewire_core
</pre>

Device Permissions
------------------
Once the kernel modules are loaded you must ensure that the user that will run the IOC application has read+write permissions on the necessary device files. The files in question can be checked for permissions with the following commands:
<pre>
ls -l /dev/fw*
crw-rw-rw-. 1 root root  248, 0 Apr 18 13:24 /dev/fw0
crw-rw-rw-. 1 root video 248, 1 Apr 18 13:24 /dev/fw1
crw-rw-rw-. 1 root root  248, 2 Apr 18 13:24 /dev/fw2
</pre>


Building the 1394 libraries
---------------------------
Please check the INSTALL or README files provided with the 1394 libraries for installation notes. Libraw1394 need to be build before libdc1394 to get the dependencies right.

The basic build and installation is fairly simple (the last step possibly as root, depending on the installation location):
<pre>
./configure
make
make install
</pre>


User Manual
===========
Diagnostic command line utility
-------------------------------
A small commandline application is provided with the firewireDCAM module. This application can be used to diagnose connection issues with a bus of many cameras - and report information about the available modes of operation for each camera found on the bus. Finally it can be used to send a reset command to each camera.

The tool comes with online help which explain the options:
<pre>
bin/linux-x86_64/firewiretool --help

./bin/linux-x86_64/firewiretool - probe firewire camera for various parameters

Usage:
./bin/linux-x86_64/firewiretool [--verbose] [--camid=0xnnnnnnnnnnnn] [--reset] [--help]
    [--report[=level]]
    --verbose: Switch on verbose print out.
    --camid:   Search for and use a camera on the bus
               with the particular ID (12 digit hex value)
    --reset:   Reset the firewire bus using the first camera
               found on each chain (generation) of the bus
    --report   Request a report of the found cameras. Will be
               limited to one camera if --camid has been used.
               The optional 'level' specifies what type of report
    --help:    This help message
</pre>

The report mode provide extensive output of all available modes:
<pre>
bin/linux-x86_64/firewiretool --report 
Enumerated 1 camera(s) on bus
_________________________________________________________________________________________
               GUID |             Vendor |                    Model |    Node |    Gen. |
_________________________________________________________________________________________
 0x00B09D01008118C9 |Point Grey Research |          Flea2 FL2-08S2C |       0 |       8 |
_________________________________________________________________________________________

    Supported modes for camera: 0x00B09D01008118C9      Flea2 FL2-08S2C
________________________________________________________________________________________
 mode |    size WxH | scalable | color coding (id) | framerates (fps)                  |
________________________________________________________________________________________
   64 |  160 x  120 |     NO   |     YUV444 ( 355) |  7.50  15.00  30.00               |
   65 |  320 x  240 |     NO   |     YUV422 ( 354) |  1.88   3.75   7.50  15.00  30.00 |
   66 |  640 x  480 |     NO   |     YUV411 ( 353) |  1.88   3.75   7.50  15.00  30.00 |
   67 |  640 x  480 |     NO   |     YUV422 ( 354) |  1.88   3.75   7.50  15.00  30.00 |
   68 |  640 x  480 |     NO   |       RGB8 ( 356) |  1.88   3.75   7.50  15.00  30.00 |
   69 |  640 x  480 |     NO   |      MONO8 ( 352) |  1.88   3.75   7.50  15.00  30.00 |
   70 |  640 x  480 |     NO   |     MONO16 ( 357) |  1.88   3.75   7.50  15.00  30.00 |
   71 |  800 x  600 |     NO   |     YUV422 ( 354) |  3.75   7.50  15.00  30.00        |
   72 |  800 x  600 |     NO   |       RGB8 ( 356) |  7.50  15.00  30.00               |
   73 |  800 x  600 |     NO   |      MONO8 ( 352) |  7.50  15.00  30.00               |
   74 | 1024 x  768 |     NO   |     YUV422 ( 354) |  1.88   3.75   7.50  15.00  30.00 |
   75 | 1024 x  768 |     NO   |       RGB8 ( 356) |  1.88   3.75   7.50  15.00        |
   76 | 1024 x  768 |     NO   |      MONO8 ( 352) |  1.88   3.75   7.50  15.00  30.00 |
   77 |  800 x  600 |     NO   |     MONO16 ( 357) |  3.75   7.50  15.00  30.00        |
   78 | 1024 x  768 |     NO   |     MONO16 ( 357) |  1.88   3.75   7.50  15.00  30.00 |
   88 | 1032 x  776 |    YES   | MONO8 YUV411 YUV422 YUV444 RGB8 MONO16 RAW8 RAW16     |
   89 |  516 x  388 |    YES   | MONO8 YUV411 YUV422 YUV444 RGB8 MONO16                |
   90 | 1032 x  388 |    YES   | MONO8 YUV411 YUV422 YUV444 RGB8 MONO16                |
________________________________________________________________________________________
</pre>


EPICS areaDetector framework
============================
The firewireDCAM module is a plug-in to the areaDetector framework. It is recommended to get familiar with the framework when starting to use this plug-in. Excellent documentation can be found on the <a href="http://cars9.uchicago.edu/software/epics/areaDetector.html">areaDetector documentation</a> page.

AreaDetector provides a number of ready-made templates, MEDM screens and plug-ins that can be used for general control and data readout. These are all fairly generic to 2D detectors. 

Note as this is a plug-in module to a larger framework, it is not considered an end-user application. A number of things will need to be done to enable an end-user to operate cameras:
* An IOC binary need to be build with a startup script and possibly substitutions files to instantiate all the necessary or desired templates.
* Custom templates or databases may need to be developed to suit specific controls requirement in your application.
* End user client display of the images is not included in this package. The data can be streamed to disk using the <a href="http://cars9.uchicago.edu/software/epics/NDPluginFile.html">NDPluginFile plug-in</a> or grabbed by any Channel Access client through the PVs available in the <a href="http://cars9.uchicago.edu/software/epics/NDPluginStdArrays.html">NDPluginStdArrays</a>

The following sections describes the templates and configuration functions this plug-in module provides and how to use them.

Database Template
-----------------
One database template is supplied to enable control and read back of all camera features. See firewireDCAM.template for details regarding macros and what features and records are available.

Startup Script
--------------
A couple of startup scripts are supplied in the example/iocBoot/iocexample/ directory. These startup scripts have a number of macros that will need to be substituted before running them. The EPICS build system at DLS will automate this and place the substituted startup files in example/bin/linux-x86/stexample.boot but other sites may possibly not have the same build system rules...

Configuration Functions
-----------------------
A number of areaDetector plug-ins are used in the example startup scripts. Please see the relevant documentation from areaDetector and modify according to your needs. The plug-ins configuration function documentation can be found on the areaDetector website: <a href="http://cars9.uchicago.edu/software/epics/NDPluginStdArrays.html#Configuration">NDPluginStdArrays</a> and <a href="http://cars9.uchicago.edu/software/epics/NDPluginROI.html#Configuration">NDPluginROI</a>

In order to initialise the firewire bus and each individual camera, two configuration functions need to be called during startup to initialise the module: 
* FDC_InitBus(): This need to be called only once to initialise the bus before FDC_Config can be called. This function does not need any arguments.
* FDC_Config(): Initialise the individual camera. 
 * Arguments: Asyn port name, camera ID (hex number), bus speed (800/400), Number of frame buffers Max memory, color mode (0 or 1)


RELEASE_NOTES.md

Master Branch (development snapshot)
------------------------------------

* Moved to building in the new areaDetector/ADCore framework.
* Removing DLS specific build dependencies.


Release 1-14
------------

* Last internal DLS release before moving to github.



firewireDCAMApp/src/firewiretool.cpp
/*
 * firewiretool.cpp
 *
 *  Created on: Mar 20, 2010
 *      Author: ulrik
 */

#include <stddef.h>
#include <stdlib.h>
#include <stddef.h>
#include <string.h>
#include <stdio.h>
#include <getopt.h>

#include <ellLib.h>


#include <dc1394/dc1394.h>
/** Number of image buffers the dc1394 library will use internally */
#define FDC_DC1394_NUM_BUFFERS 5
/** Print an errorcode to stderr. */
#define ERR(errCode) \
    do { \
        int __err__ = (errCode); \
        if (__err__ != 0) \
            fprintf(stderr, \
                "ERROR [%s:%d]: dc1394 code: %d\n", \
                __FILE__, __LINE__, __err__); \
    } while (0)

typedef struct {
    dc1394color_coding_t code;
    const char * colorCodeStr;
} dc1394_colorCodeString_t;
static dc1394_colorCodeString_t colorCodeString_lut[] = {
    {DC1394_COLOR_CODING_MONO8,   "MONO8"},
    {DC1394_COLOR_CODING_YUV411,  "YUV411"},
    {DC1394_COLOR_CODING_YUV422,  "YUV422"},
    {DC1394_COLOR_CODING_YUV444,  "YUV444"},
    {DC1394_COLOR_CODING_RGB8,    "RGB8"},
    {DC1394_COLOR_CODING_MONO16,  "MONO16"},
    {DC1394_COLOR_CODING_RGB16,   "RGB16"},
    {DC1394_COLOR_CODING_MONO16S, "MONO16S"},
    {DC1394_COLOR_CODING_RGB16S,  "RGB16S"},
    {DC1394_COLOR_CODING_RAW8,    "RAW8"},
    {DC1394_COLOR_CODING_RAW16,   "RAW16"},
};

struct camLinkedList {
    camLinkedList *next;
    dc1394camera_t *camera;
};
typedef struct camLinkedList camListItem_t;

static struct option longopts[] = {
        {"verbose",         no_argument,        NULL,   0},
        {"camid",           required_argument,  NULL,   1},
        {"reset",           no_argument,        NULL,   2},
        {"report",          optional_argument,  NULL,   3},
        {"help",            no_argument,        NULL,   4},
        { NULL,             0,                  NULL,   0}
};

struct fwtooloptions_t {
    int verbose;
    int reset;
    int report;
    char *camid;
};
static struct fwtooloptions_t fwtopts = { 0, 0, 0, NULL };
void parseOptions(int argc, char* argv[]);

dc1394camera_t* findcam(dc1394camera_t **cams, int ncams, char* guid);
int allocate_cams(dc1394_t *bus, dc1394camera_list_t *list, dc1394camera_t **cams, int verbose);
void reportModes(dc1394camera_t *cam);
void reportFeatures(dc1394camera_t *cam);
const char* getColorCodeString(dc1394color_coding_t code);
void reset_bus();

void parseOptions(int argc, char* argv[])
{
    int c;
    while( (c = getopt_long( argc, argv, "", longopts, NULL)) != -1)
    {
        switch(c)
        {
        case 0:
            fwtopts.verbose = 1;
            break;
        case 1:
            fwtopts.camid = strdup(optarg);
            break;
        case 2:
            fwtopts.reset = 1;
            break;
        case 3:
            fwtopts.report = 1;
            if (optarg!=NULL)
                fwtopts.report = atoi(optarg);
            break;
        default:
            printf( "\n"
                    "%s - probe firewire camera for various parameters\n\n"
                    "Usage:\n"
                    "%s [--verbose] [--camid=0xnnnnnnnnnnnn] [--reset] [--help]\n"
                    "    [--report[=level]]\n"
                    "    --verbose: Switch on verbose print out.\n"
                    "    --camid:   Search for and use a camera on the bus\n"
                    "               with the particular ID (12 digit hex value)\n"
                    "    --reset:   Reset the firewire bus using the first camera\n"
                    "               found on each chain (generation) of the bus\n"
                    "    --report   Request a report of the found cameras. Will be\n"
                    "               limited to one camera if --camid has been used.\n"
                    "               The optional \'level\' specifies what type of report\n"
                    "    --help:    This help message\n"
                    "\n", argv[0], argv[0]);
            exit(0);
            break;
        }
    }
    if(fwtopts.verbose==1)
    {
        printf("\nOption parsed:\n"
                "    camid = %s\n"
                "    reset = %d\n"
                "    report = %d\n"
                ,fwtopts.camid, fwtopts.reset, fwtopts.report);
    }
}

int main(int argc,char *argv[])
{

    dc1394_t * bus;
    dc1394camera_list_t * list;
    dc1394camera_t ** cams;
    dc1394camera_t ** tmpcams;
    dc1394camera_t * cam = NULL;
    dc1394camera_t *tmpcam = NULL;
/*    camListItem_t *head, *curr;*/
    dc1394error_t err;
    int i, nReports=0;
    int nAllocatedCams = 0;

    // First parse the options the user requested
    parseOptions(argc, argv);

    // Allocate a handle to the bus
    bus = dc1394_new ();
    if (bus == NULL)
    {
        fprintf(stderr, "ERROR [%s:%d]: could not create dc1394 context.\n", __FILE__, __LINE__);
        return -1;
    }

    // Enumerate all the cameras on the bus (gives us a list of available cameras)
    err=dc1394_camera_enumerate (bus, &list);
    if (err!=0)
    {
        ERR( err );
        dc1394_free (bus);
        return -2;
    }
    printf("Enumerated %d camera(s) on bus\n", list->num);
    cams = (dc1394camera_t **)calloc(list->num + 1, sizeof(dc1394camera_t *));

    // Allocate memory and handle to all the cameras on the bus
    nAllocatedCams = allocate_cams(bus, list, cams, 1);
    if (nAllocatedCams < 1)
    {
        fprintf(stderr, "ERROR [%s:%d] Did not manage to allcoate a single camera\n", __FILE__, __LINE__);
        return -3;
    }

    // Create a linked list with all the cameras we are interested in
    // (currently either all or the specified --camid)
/*  if (fwtopts.camid != NULL)
    {
        cam = findcam(cams, nAllocatedCams, fwtopts.camid);
        if (cam == NULL) return -4;
        head = (camListItem_t*)calloc(1,sizeof(camListItem_t));
        head->camera = cam;
        head->next = NULL;
    } else
    {
        for (i=0; i < nAllocatedCams; i++)
        {
            curr = (camListItem_t*)calloc(1,sizeof(camListItem_t));
            curr->camera =
        }
    }
*/
    if (fwtopts.report > 0)
    {
        if (cam == NULL)
        {
            nReports = nAllocatedCams;
            tmpcams = cams;
        } else
        {
            nReports = 1;
            tmpcams = &cam;
        }

        for (i=0; i<nReports; i++)
        {
            tmpcam = *(tmpcams+i);
            //ERR( dc1394_camera_print_info(tmpcam, stdout) );
            reportModes(tmpcam);
            reportFeatures(tmpcam);
        }

    }

    if (fwtopts.reset) reset_bus();

    dc1394_camera_free_list (list);
    dc1394_free (bus);

    return(0);
}

int allocate_cams(dc1394_t *bus, dc1394camera_list_t *list, dc1394camera_t **cams, int verbose)
{
    unsigned int i, cnt=0;
    dc1394camera_t* camera;
    uint32_t node, generation;
    unsigned int tblWidth = 89;

    for (i=0;i<tblWidth;i++) printf("_");
    printf( "\n");
    printf("%19s |%19s |%25s |%8s |%8s |\n", "GUID", "Vendor", "Model","Node","Gen.");
    for (i=0;i<tblWidth;i++) printf("_");
    printf( "\n");

    // run through all the identified cameras and allocate them
    for (i = 0; i < list->num; i++)
    {
        camera = dc1394_camera_new(bus, list->ids[i].guid);
        if (camera != NULL)
        {
            ERR( dc1394_camera_get_node(camera, &node, &generation) );
            if (verbose > 0)
                printf(" 0x%16.16lX |%19s |%25s |%8d |%8d |\n",
                        camera->guid, camera->vendor, camera->model, node, generation);
            cams[cnt] = camera;
            cnt++;
        }
    }
    for (i=0;i<tblWidth;i++) printf("_");
    printf( "\n");
    return cnt;
}
typedef struct camNode_t {
    ELLNODE node;
    uint32_t generation;
    dc1394camera_t *cam;
}camNode_t;


void reset_bus()
{
    dc1394_t * d;
    dc1394camera_list_t * list;
    dc1394camera_t *cam = NULL;
    uint32_t generation, latch=0;
    uint32_t node;
    ELLLIST camList;
    camNode_t *camListItem, *tmp;
    unsigned int i, newBus;

    d = dc1394_new ();
    ERR( dc1394_camera_enumerate (d, &list) );

    // To reset a multi-bus system it is necessary to find a camera on each
    // individual bus and call the reset function with that camera.
    ellInit(&camList);

    // Get the 'generation' parameter for each camera. This parameter indicate
    // which bus the camera is located on.
    // For each specific 'generation' we add the camera handle to a list which
    // we can later use to reset each bus.
    for (i=0;i<list->num; i++)
    {
        fflush(stdout);
        cam = dc1394_camera_new (d, list->ids[i].guid);
        ERR( dc1394_camera_get_node(cam, &node, &generation) );

        // Run through the collected list of cameras and check if anyone
        // has the same 'generation' parameter... (i.e. is on the same bus)
        tmp=(camNode_t*)ellFirst(&camList);
        newBus = 1;
        while(tmp!=NULL)
        {
            if (generation == tmp->generation)
            {
                newBus = 0;
                break;
            }
            tmp=(camNode_t*)ellNext((ELLNODE*)tmp);
        }

        // If we havent already listed a camera on this bus -or if this is the
        // first camera we check: add the camera handle to a list of cameras that
        // we want to use for resetting busses.
        // Else free up the camera handle as we won't use it until we instantiate
        // our driver plugin.
        if (newBus==1 || i==0)
        {
            camListItem = (camNode_t*)calloc(1, sizeof(camNode_t));
            camListItem->cam = cam;
            camListItem->generation = generation;
            ellAdd(&camList, (ELLNODE*)camListItem);
            latch = generation;
        } else
        {
            // if we dont need the camera handle to reset the bus
            // we might as well free it up
            dc1394_camera_free(cam);
        }
    }

    // Go through the list of cameras that have been identified to be
    // on separate physical busses. Call reset for each of them and free
    // up the camera handle
    camListItem = (camNode_t*)ellFirst(&camList);
    while(camListItem != NULL)
    {
        printf("Resetting bus: %3d using cam: 0x%16.16lX... ", camListItem->generation, camListItem->cam->guid);
        fflush(stdout);
        ERR( dc1394_reset_bus(camListItem->cam) );
        printf("Done\n");
        dc1394_camera_free(camListItem->cam);
        camListItem = (camNode_t*)ellNext((ELLNODE*)camListItem);
    }

    // Clear up after ourselves.
    ellFree(&camList);
    dc1394_camera_free_list (list);
    dc1394_free (d);
    printf("\n");
    return;
}  

dc1394camera_t* findcam(dc1394camera_t **cams, int ncams, char* guid)
{
    int i;
    uint64_t iguid = 0;
    unsigned int ret;

    // First parse the GUI string into a 64bit integer if possible
    if (guid != NULL)
    {
        ret = sscanf(guid, "0x%16lX", &iguid);
        if (ret != 1)
        {
            fprintf(stderr, "ERROR [%s:%d] could not parse GUID string: \"%s\"\n", __FILE__, __LINE__, guid);
            return NULL;
        }
    }

    // run through all the identified cameras on the bus and check if the
    // ID matches the requested one
    for (i = 0; i < ncams; i++)
    {
        if (cams[i]->guid == iguid) return cams[i];
    }

    fprintf(stderr, "Could not find requested camera: %s\n", guid);
    return NULL;
}


void reportModes(dc1394camera_t *cam)
{
    dc1394video_modes_t video_modes;
    dc1394framerates_t framerates;
    dc1394color_coding_t color_coding;
    dc1394color_codings_t codings;
    dc1394bool_t is_scalable;
    dc1394video_mode_t mode;
    uint32_t width, height;
    float framerate;
    unsigned int i,j,q;
    unsigned int tblWidth = 88;

    ERR( dc1394_video_get_supported_modes(cam, &video_modes) );

    printf("\n    Supported modes for camera: 0x%16.16lX %20s\n", cam->guid, cam->model);
    for (i=0;i<tblWidth;i++) printf("_");
    printf( "\n");

    printf("%5s |","mode");
    printf("%12s |","size WxH");
    printf(" scalable |");
    printf(" %15s |", "color coding (id)");
    printf(" %-33s |\n", "framerates (fps)");
    for (i=0;i<tblWidth;i++) printf("_");
    printf( "\n");
    for(i=0; i<video_modes.num; i++)
    {
        mode = video_modes.modes[i];
        ERR( dc1394_get_image_size_from_video_mode(cam, mode, &width, &height) );
        is_scalable = dc1394_is_video_mode_scalable (mode);
        ERR( dc1394_get_color_coding_from_video_mode(cam, mode, &color_coding) );
        if (is_scalable)
        {
            ERR( dc1394_format7_get_max_image_size (cam, mode, &width, &height) );
            ERR( dc1394_format7_get_color_codings (cam,mode, &codings) );
        }

        printf(" %4d |", (unsigned int)mode);
        printf(" %4d x %4d |", width, height);
        is_scalable ? printf("    YES   |") : printf("     NO   |");

        if (!is_scalable)
        {
            printf(" %10s (%4d) |", getColorCodeString(color_coding), (unsigned int)color_coding);
            ERR( dc1394_video_get_supported_framerates(cam, mode, &framerates) );
            for (j=0; j<framerates.num; j++)
            {
                ERR( dc1394_framerate_as_float(framerates.framerates[j], &framerate) );
                printf(" %5.2f ", framerate);
            }
            for (q=0;q<(5-framerates.num);q++) printf("       ");
        } else
        {
            int len = 0;
            for (j=0; j<codings.num; j++)
            {
                len += printf(" %s", getColorCodeString(codings.codings[j]));
            }
            for (q=len;q<55;q++) printf(" ");
        }
        printf("|\n");

    }
    for (i=0;i<tblWidth;i++) printf("_");
    printf( "\n");
}

void reportFeatures(dc1394camera_t *cam)
{
    dc1394featureset_t features;
    ERR( dc1394_feature_get_all (cam, &features) );
    //ERR( dc1394_feature_print_all (&features, stdout) );
}


const char* getColorCodeString(dc1394color_coding_t code)
{
    unsigned int i;

    for (i=0; i<DC1394_COLOR_CODING_NUM; i++)
        if (colorCodeString_lut[i].code == code) return colorCodeString_lut[i].colorCodeStr;
    return "";
}
